benchadapt/python/README.md

```diff
@@ -1,3 +1,129 @@
+
+The `benchadapt.adapters.BenchmarkAdapter` abstract class defines a basic workflow:
+
+1. Call an arbitrary `command` shell command, presumably to run benchmarks. If results
+are already guaranteed to exist, this can be set to do nothing.
+2. Transform results produced by the benchmarks into a list of `BenchmarkResult` instances.
+3. Postprocess results to ensure a consistent `run_id` and override any metadata fields
+not already set correctly.
+4. Post each result to a Conbench server.
+
+Classes that inherit from the abstract class need to define
+
+1. How to get results, including what `command` should be (though it can be defined later
+by the user, if desired) and how to get the raw results (e.g. if they are in a file or
+directory of files, where they are and how to read them in).
+2. How to transform the results into a list of `BenchmarkResult` instances ((2) above) in
+the `._transform_results()` method.
+
+(3) and (4) are handled by the abstract class.
+
+Various adapters are alrady defined in the `adapters` submodule, including ones for
+Google Benchmark and Folly, as well as a generic `CallableAdapter`, which takes a Python
+`Callable` instance (a function or class with a `__call__()` method) that returns a list
+of `BenchmarkResult` instances directly instead of a shell command. Many more adapters
+are possible; if you create one corresponding to a benchmarking tool, please make a PR!
+
+#### Running an adapter
+
+Adapters have separate `.run()` and `.publish_results()` methods; the former runs the
+benchmarks, transforms the results, and stores them in a `.results` attribute of the
+instance. It does not post them, so is useful for looking at results interactively before
+sending them. `.publish_results()` takes the results from the `.results` attribute and
+posts them to a Conbench server.
+
+The whole instance also has a `__call__()` method defined so it can be called like a
+function that both runs and publishes, so a somewhat minimal script for running
+benchmarks in CI might look like
+
+``` python
+import os
+
+from benchadapt.adapters import GoogleBenchmarkAdapter
+
+adapter = GoogleBenchmarkAdapter(
+    command=["bash", "./run-benchmarks.sh"],
+    result_file="benchmarks.json",
+    result_fields_override={
+        "run_reason": os.getenv("CONBENCH_RUN_REASON")
+    },
+    result_fields_append={
+        "info": {"build_version": os.getenv("MY_BUILD_VERSION")}
+    }
+)
+adapter()
+```
+
+Of note:
+
+- `result_fields_override` will replace the whole attribute with a new value. This works
+with all types (strings, dicts, etc.), so here `run_reason` will be set for all results.
+- `result_fields_append` will append the new values to dicts which may already have data.
+Here, `build_version` will be appended to the `info` dict. (In this case it is empty
+anyway, so this is equivalent to
+`result_fields_override={"info": {"build_version": os.getenv("MY_BUILD_VERSION")}})`.)
```

Comment:
It might be nice to have (another?) example where `result_fields_append` does actually append

Reply:
Yeah, I was trying to figure out a good example here. `tags` is populated, but for the most part people shouldn't be messing with that at this level. `context` might be a better example; `"benchmark_language"` will already be set, but we could append `"compiler_flags"` or something

Reply:
Left this one (I think seeing how they connect is helpful), but added a new `context` one to make a use-case more obvious

Reply:
Awesome, yeah that sounds good